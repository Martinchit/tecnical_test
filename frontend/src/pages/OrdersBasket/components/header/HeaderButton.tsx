import styled from 'styled-components';
import { Button } from 'react-bootstrap';

export const HeaderButton = styled(Button)`
  margin-left: 2%;
  width: 10%;

  @media (max-width: 767px) {
    width: 45%;
    margin-top: 3%;
    font-size: 0.8rem;
  }

  &:focus {
    outline: none;
    box-shadow: none;
  }
`;
