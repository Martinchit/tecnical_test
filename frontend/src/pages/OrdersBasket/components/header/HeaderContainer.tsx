import styled from 'styled-components';
import { FlexBox } from '../../../../core/components/atoms/FlexBox';

export const HeaderContainer = styled(FlexBox)`
  justify-content: flex-end;
  margin: 2% 0;
  width: 100%;

  @media (max-width: 767px) {
    justify-content: center;
  }
`;
