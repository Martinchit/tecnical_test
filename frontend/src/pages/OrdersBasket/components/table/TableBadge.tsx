import React from 'react';
import styled from 'styled-components';
import { Badge as B } from 'react-bootstrap';
import { OrderStatus } from '../../../../types/types';

interface TableBadge {
  orderStatus: OrderStatus;
}

export const TableBadge: React.FC<TableBadge> = ({ orderStatus }) => {
  const getVariant = () => {
    switch (orderStatus) {
      case 'Booked':
        return 'success';
      case 'In Progress':
        return 'info';
      case 'Rejected':
        return 'danger';
      case 'Ready':
        return 'primary';
      default:
        return 'warning';
    }
  };
  return <Badge variant={getVariant()}>{orderStatus}</Badge>;
};

const Badge = styled(B)`
  width: 80%;
  @media (max-width: 767px) {
    width: 100%;
  }
`;
