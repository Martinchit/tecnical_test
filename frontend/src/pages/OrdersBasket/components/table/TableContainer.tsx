import styled from 'styled-components';
import { Table } from 'react-bootstrap';

export const TableContainer = styled(Table)`
  padding-bottom: 100px;
`;
