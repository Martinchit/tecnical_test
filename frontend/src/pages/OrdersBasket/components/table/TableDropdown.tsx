import React from 'react';
import { Dropdown } from 'react-bootstrap';
import { TableDropdownButton } from './TableDropdownButton';
import { OrderExecutionMode } from '../../../../types/types';
import { ExecutionMode } from '../../../../core/lib/utils/executionMode';

interface TableDropdownProps {
  title: OrderExecutionMode | undefined;
  dropdownID: string;
  onSelect: (id: number) => void;
  disabled: boolean;
}

export const TableDropdown: React.FC<TableDropdownProps> = ({
  title,
  dropdownID,
  onSelect,
  disabled,
}) => (
  <Dropdown onSelect={(id: string) => onSelect(Number(id))}>
    <TableDropdownButton
      className="pull-right"
      disabled={disabled}
      variant="light"
      id={dropdownID}
    >
      {title}
    </TableDropdownButton>
    <Dropdown.Menu>
      {ExecutionMode.map((option, i) => (
        <Dropdown.Item eventKey={`${i}`} key={option}>
          {option}
        </Dropdown.Item>
      ))}
    </Dropdown.Menu>
  </Dropdown>
);
