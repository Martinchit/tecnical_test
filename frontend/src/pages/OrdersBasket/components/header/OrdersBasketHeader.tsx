import React from 'react';
import { HeaderContainer } from './HeaderContainer';
import { HeaderButton } from './HeaderButton';

interface OrdersBasketHeaderProps {
  onRemoveButtonClickHandler: () => void;
  onBookButtonClickHandler: () => void;
  removeButtonDisable: boolean;
  bookButtonDisable: boolean;
}

export const OrdersBasketHeader: React.FC<OrdersBasketHeaderProps> = ({
  onRemoveButtonClickHandler,
  onBookButtonClickHandler,
  removeButtonDisable,
  bookButtonDisable,
}) => {
  return (
    <HeaderContainer>
      <HeaderButton
        disabled={removeButtonDisable}
        variant="primary"
        onClick={onRemoveButtonClickHandler}
      >
        Remove
      </HeaderButton>
      <HeaderButton
        disabled={bookButtonDisable}
        variant="primary"
        onClick={onBookButtonClickHandler}
      >
        Book
      </HeaderButton>
    </HeaderContainer>
  );
};
