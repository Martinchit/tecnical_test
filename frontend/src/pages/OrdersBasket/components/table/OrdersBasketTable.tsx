import React from 'react';
import { Form, InputGroup, Button } from 'react-bootstrap';
import { TableContainer } from './TableContainer';
import { TableBadge } from './TableBadge';
import { TableDropdown } from './TableDropdown';
import { TableInputField } from './TableInputField';
import { FlexBox } from '../../../../core/components/atoms/FlexBox';
import { TableErrorContainer } from '../../../../core/components/atoms/TableErrorContainer';
import { TableCell } from '../../../../core/components/atoms/TableCell';
import { TableHead } from '../../../../core/components/atoms/TableHead';
import { Alert } from '../../../../core/components/atoms/Alert';
import { StockOrder } from '../../../../types/types';

interface OrdersBasketTableProps {
  orders: StockOrder[];
  onSelectCheckClickHandler: (orderId: string, selected: boolean) => void;
  onExecutionModeSelectHandler: (stockIdx: number, modeIdx: number) => void;
  onOrderPriceChangeHandler: (stockIdx: number, orderPrice: string) => void;
  onShareAmountChangeHandler: (stockIdx: number, shareAmount: string) => void;
  onCancelButtonClickHandler: (stockIdx: number) => void;
}

export const OrdersBasketTable: React.FC<OrdersBasketTableProps> = ({
  onSelectCheckClickHandler,
  onExecutionModeSelectHandler,
  onOrderPriceChangeHandler,
  onShareAmountChangeHandler,
  onCancelButtonClickHandler,
  orders,
}) => {
  return (
    <TableContainer responsive>
      <thead>
        <tr>
          <TableHead />
          <TableHead>Status</TableHead>
          <TableHead>Side</TableHead>
          <TableHead>Stock Code</TableHead>
          <TableHead>Execution Code</TableHead>
          <TableHead>Order Price</TableHead>
          <TableHead>Amount &#40;Shares&#41;</TableHead>
        </tr>
      </thead>
      <tbody>
        {orders.map(
          (
            {
              id,
              side,
              bloombergTickerLocal,
              status,
              executionMode,
              currency,
              orderPrice,
              shareAmount,
              error,
              selected,
              orderId,
            },
            i
          ) => {
            const disabledOnMarketMode =
              executionMode !== undefined && executionMode === 'Market';
            const disabledOnSubmittedCondition =
              status !== 'Not Ready' && status !== 'Ready';
            return (
              <React.Fragment key={orderId}>
                <tr>
                  <TableCell>
                    <Form.Check
                      checked={selected}
                      onChange={() => onSelectCheckClickHandler(orderId, selected)}
                    />
                  </TableCell>
                  <TableCell>
                    <TableBadge orderStatus={status} />
                  </TableCell>
                  <TableCell>{side}</TableCell>
                  <TableCell>{bloombergTickerLocal}</TableCell>
                  <TableCell>
                    <TableDropdown
                      dropdownID={`${id}_${side}_mode_dropdown`}
                      title={executionMode}
                      disabled={disabledOnSubmittedCondition}
                      onSelect={(optionId) =>
                        onExecutionModeSelectHandler(i, Number(optionId))
                      }
                    />
                  </TableCell>
                  <TableCell>
                    <InputGroup>
                      <TableInputField
                        disabled={
                          disabledOnMarketMode || disabledOnSubmittedCondition
                        }
                        value={orderPrice || ''}
                        type="number"
                        aria-describedby={`${id}_${side}_input_append`}
                        onChange={({ target }) =>
                          onOrderPriceChangeHandler(i, target.value)
                        }
                      />
                      <InputGroup.Append id={`${id}_${side}_input_append`}>
                        <InputGroup.Text>{currency}</InputGroup.Text>
                      </InputGroup.Append>
                    </InputGroup>
                  </TableCell>
                  <TableCell>
                    <TableInputField
                      disabled={disabledOnSubmittedCondition}
                      value={shareAmount || ''}
                      type="number"
                      onChange={({ target }) =>
                        onShareAmountChangeHandler(i, target.value)
                      }
                    />
                  </TableCell>
                </tr>
                {error ? (
                  <tr>
                    <TableErrorContainer colSpan={7}>
                      <Alert variant="danger">
                        <FlexBox justifyContent="space-between">
                          {`Error: ${error}`}
                          <Button
                            variant="light"
                            onClick={() => onCancelButtonClickHandler(i)}
                          >
                            Close
                          </Button>
                        </FlexBox>
                      </Alert>
                    </TableErrorContainer>
                  </tr>
                ) : null}
              </React.Fragment>
            );
          }
        )}
      </tbody>
    </TableContainer>
  );
};
