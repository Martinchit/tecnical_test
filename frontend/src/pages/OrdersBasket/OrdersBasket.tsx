import React from 'react';
import { OrdersBasketHeader } from './components/header/OrdersBasketHeader';
import { OrdersBasketTable } from './components/table/OrdersBasketTable';
import { PageTitle } from '../../core/components/molecules/PageTitle';
import { FlexBox } from '../../core/components/atoms/FlexBox';
import { Text } from '../../core/components/atoms/Text';
import { StockOrder, OrderExecutionMode, OrderStatus } from '../../types/types';
import { ExecutionMode } from '../../core/lib/utils/executionMode';

interface OrdersBasketState {
  loading: boolean;
}

interface OrdersBasketProps {
  removeStockInOrderAction: () => void;
  selectStockInOrderAction: (orderId: string) => void;
  unselectStockInOrderAction: (orderId: string) => void;
  updateStockOrderExecutionModeAction: (
    stockIdx: number,
    executionMode: OrderExecutionMode
  ) => void;
  updateStockOrderOrderPriceAction: (
    stockIdx: number,
    orderPrice: number
  ) => void;
  updateStockOrderShareAmountAction: (
    stockIdx: number,
    shareAmount: number
  ) => void;
  updateStockOrderStatusAction: (stockIdx: number, status: OrderStatus) => void;
  placeStockOrderAction: (stockOrder: StockOrder, token: string) => void;
  acknowledgeStockOrderErrorAction: (stockIdx: number) => void;
  logOutAction: () => void;
  orders: StockOrder[];
  token: string;
  sessionTimeout: boolean;
}

export class OrdersBasket extends React.Component<
OrdersBasketProps,
OrdersBasketState
> {
  componentDidUpdate = () => {
    const { sessionTimeout, logOutAction } = this.props;
    if (sessionTimeout) {
      alert('Session Timeout - Please log in again!');
      logOutAction();
    }
  };

  onRemoveButtonClickHandler = () => {
    const { removeStockInOrderAction } = this.props;
    removeStockInOrderAction();
  };

  onBookButtonClickHandler = () => {
    const { orders, placeStockOrderAction, token } = this.props;
    const ordersToBePlaced = orders.filter(({ selected }) => selected);
    for (const order of ordersToBePlaced) {
      placeStockOrderAction(order, token);
    }
  };

  onSelectCheckClickHandler = (orderId: string, selected: boolean) => {
    const { selectStockInOrderAction, unselectStockInOrderAction } = this.props;
    if (selected) {
      unselectStockInOrderAction(orderId);
    } else {
      selectStockInOrderAction(orderId);
    }
  };

  onExecutionModeSelectHandler = (stockIdx: number, modeIdx: number) => {
    const { orders, updateStockOrderExecutionModeAction } = this.props;
    const {
      orderPrice,
      shareAmount,
      status,
      currency,
      bloombergTickerLocal,
    } = orders[stockIdx];
    const executionMode = ExecutionMode[modeIdx];
    updateStockOrderExecutionModeAction(stockIdx, executionMode);
    if (executionMode === 'Market') {
      const { updateStockOrderOrderPriceAction } = this.props;
      updateStockOrderOrderPriceAction(stockIdx, 0);
    }
    if (currency && bloombergTickerLocal) {
      this.checkOrderStatus(
        executionMode,
        shareAmount,
        orderPrice,
        status,
        stockIdx
      );
    }
  };

  onOrderPriceChangeHandler = (stockIdx: number, orderPrice: string) => {
    const { orders, updateStockOrderOrderPriceAction } = this.props;
    const {
      executionMode,
      shareAmount,
      status,
      currency,
      bloombergTickerLocal,
    } = orders[stockIdx];
    const converted = Number(orderPrice);
    updateStockOrderOrderPriceAction(stockIdx, converted);
    if (currency && bloombergTickerLocal) {
      this.checkOrderStatus(
        executionMode,
        shareAmount,
        converted,
        status,
        stockIdx
      );
    }
  };

  onShareAmountChangeHandler = (stockIdx: number, shareAmount: string) => {
    const { orders, updateStockOrderShareAmountAction } = this.props;
    const {
      executionMode,
      orderPrice,
      status,
      currency,
      bloombergTickerLocal,
    } = orders[stockIdx];
    const converted = Number(shareAmount);
    updateStockOrderShareAmountAction(stockIdx, converted);
    if (currency && bloombergTickerLocal) {
      this.checkOrderStatus(
        executionMode,
        converted,
        orderPrice,
        status,
        stockIdx
      );
    }
  };

  checkOrderStatus = (
    eM: OrderExecutionMode | undefined,
    sA: number | undefined,
    oP: number | undefined,
    s: OrderStatus,
    id: number
  ) => {
    const validSA = this.isValidShareAmount(sA);
    const validOP = this.isValidOrderPrice(oP);
    if (s === 'Not Ready') {
      if (
        (eM === 'Market' && validSA) ||
        (eM === 'Limit' && validSA && validOP)
      ) {
        this.updateOrderStatus(id, 'Ready');
      }
    }
    if (s === 'Ready') {
      if (!validSA || (eM === 'Limit' && !validOP)) {
        this.updateOrderStatus(id, 'Not Ready');
      }
    }
  };

  isValidShareAmount = (a: number | undefined) => a && a > 0;

  isValidOrderPrice = (p: number | undefined) => p && p > 0;

  updateOrderStatus = (idx: number, status: OrderStatus) => {
    const { updateStockOrderStatusAction } = this.props;
    updateStockOrderStatusAction(idx, status);
  };

  onCancelButtonClickHandler = (stockIdx: number) => {
    const { acknowledgeStockOrderErrorAction } = this.props;
    acknowledgeStockOrderErrorAction(stockIdx);
  };

  render() {
    const { orders } = this.props;
    const selectedOrders = orders.filter(({ selected }) => selected);
    const isSelectedOrdersEmpty = selectedOrders.length === 0;
    const isInvalidOrderExisted =
      selectedOrders.filter(
        ({ status }) =>
          status === 'Not Ready' || status === 'Booked' || status === 'Rejected'
      ).length > 0;
    return (
      <FlexBox>
        <PageTitle title="Orders Basket" />
        <OrdersBasketHeader
          onRemoveButtonClickHandler={this.onRemoveButtonClickHandler}
          onBookButtonClickHandler={this.onBookButtonClickHandler}
          removeButtonDisable={isSelectedOrdersEmpty}
          bookButtonDisable={isSelectedOrdersEmpty || isInvalidOrderExisted}
        />
        <FlexBox margin="2% 0%">
          {orders.length ? (
            <OrdersBasketTable
              orders={orders}
              onSelectCheckClickHandler={this.onSelectCheckClickHandler}
              onExecutionModeSelectHandler={this.onExecutionModeSelectHandler}
              onOrderPriceChangeHandler={this.onOrderPriceChangeHandler}
              onShareAmountChangeHandler={this.onShareAmountChangeHandler}
              onCancelButtonClickHandler={this.onCancelButtonClickHandler}
            />
          ) : (
            <Text>No Stock Order is selected in the basket</Text>
          )}
        </FlexBox>
      </FlexBox>
    );
  }
}
