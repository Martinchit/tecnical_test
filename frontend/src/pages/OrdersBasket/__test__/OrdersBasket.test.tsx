import React from 'react';
import { shallow } from 'enzyme';

import { OrdersBasket } from '../OrdersBasket';
import { OrdersBasketHeader } from '../components/header/OrdersBasketHeader';
import { OrdersBasketTable } from '../components/table/OrdersBasketTable';
import {
  OrderSide,
  OrderStatus,
  OrderExecutionMode,
} from '../../../types/types';

describe('OrdersBasket', () => {
  const orderId = '9cc4ca45-b42b-462d-a8aa-931e6041cfcb';
  const props = {
    removeStockInOrderAction: jest.fn(),
    selectStockInOrderAction: jest.fn(),
    unselectStockInOrderAction: jest.fn(),
    updateStockOrderExecutionModeAction: jest.fn(),
    updateStockOrderOrderPriceAction: jest.fn(),
    updateStockOrderShareAmountAction: jest.fn(),
    updateStockOrderStatusAction: jest.fn(),
    placeStockOrderAction: jest.fn(),
    acknowledgeStockOrderErrorAction: jest.fn(),
    logOutAction: jest.fn(),
    orders: [
      {
        id: 1,
        stockId: 'testID',
        currency: 'HKD',
        bloombergTickerLocal: 'testTicker',
        side: 'Buy' as OrderSide,
        status: 'Not Ready' as OrderStatus,
        selected: false,
        orderId,
      },
    ],
    token: 'testToken',
    sessionTimeout: false,
  };

  const restoreMock = () => {
    props.removeStockInOrderAction.mockRestore();
    props.selectStockInOrderAction.mockRestore();
    props.unselectStockInOrderAction.mockRestore();
    props.updateStockOrderExecutionModeAction.mockRestore();
    props.updateStockOrderOrderPriceAction.mockRestore();
    props.updateStockOrderShareAmountAction.mockRestore();
    props.updateStockOrderStatusAction.mockRestore();
    props.placeStockOrderAction.mockRestore();
    props.acknowledgeStockOrderErrorAction.mockRestore();
  };

  it('renders correctly', () => {
    const wrapper = shallow(<OrdersBasket {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles onRemoveButtonClickHandler in OrdersBasketHeader', () => {
    const wrapper = shallow(<OrdersBasket {...props} />);
    wrapper
      .find(OrdersBasketHeader)
      .first()
      .props()
      .onRemoveButtonClickHandler();
    expect(props.removeStockInOrderAction).toHaveBeenCalled();
    restoreMock();
  });

  it('handles onBookButtonClickHandler in OrdersBasketHeader', () => {
    const p = { ...props, orders: [{ ...props.orders[0], selected: true }] };
    const wrapper = shallow(<OrdersBasket {...p} />);
    wrapper.find(OrdersBasketHeader).first().props().onBookButtonClickHandler();
    expect(props.placeStockOrderAction).toHaveBeenCalled();
    expect(props.placeStockOrderAction).toHaveBeenCalledTimes(1);
    expect(props.placeStockOrderAction).toHaveBeenCalledWith(
      p.orders[0],
      props.token
    );
    restoreMock();
  });

  it('handles onSelectCheckClickHandler in OrdersBasketTable & calls unselectStockInOrderAction', () => {
    const selected = true;
    const wrapper = shallow(<OrdersBasket {...props} />);
    wrapper
      .find(OrdersBasketTable)
      .first()
      .props()
      .onSelectCheckClickHandler(orderId, selected);
    expect(props.unselectStockInOrderAction).toHaveBeenCalled();
    expect(props.unselectStockInOrderAction).toHaveBeenCalledWith(orderId);
    restoreMock();
  });

  it('handles onSelectCheckClickHandler in OrdersBasketTable & calls selectStockInOrderAction', () => {
    const selected = false;
    const wrapper = shallow(<OrdersBasket {...props} />);
    wrapper
      .find(OrdersBasketTable)
      .first()
      .props()
      .onSelectCheckClickHandler(orderId, selected);
    expect(props.selectStockInOrderAction).toHaveBeenCalled();
    expect(props.selectStockInOrderAction).toHaveBeenCalledWith(orderId);
    restoreMock();
  });

  it('handles onExecutionModeSelectHandler in OrdersBasketHeader', () => {
    const stockIdx = 0;
    const modeId = 0;
    const wrapper = shallow(<OrdersBasket {...props} />);
    wrapper
      .find(OrdersBasketTable)
      .first()
      .props()
      .onExecutionModeSelectHandler(stockIdx, modeId);
    expect(props.updateStockOrderExecutionModeAction).toHaveBeenCalled();
    expect(props.updateStockOrderExecutionModeAction).toHaveBeenCalledWith(
      stockIdx,
      'Market'
    );
    expect(props.updateStockOrderOrderPriceAction).toHaveBeenCalled();
    expect(props.updateStockOrderOrderPriceAction).toHaveBeenCalledWith(
      stockIdx,
      0
    );
    restoreMock();
  });

  it('handles onOrderPriceChangeHandler in OrdersBasketHeader', () => {
    const stockIdx = 0;
    const orderPrice = '100';
    const wrapper = shallow(<OrdersBasket {...props} />);
    wrapper
      .find(OrdersBasketTable)
      .first()
      .props()
      .onOrderPriceChangeHandler(stockIdx, orderPrice);
    expect(props.updateStockOrderOrderPriceAction).toHaveBeenCalled();
    expect(props.updateStockOrderOrderPriceAction).toHaveBeenCalledWith(
      stockIdx,
      Number(orderPrice)
    );
    restoreMock();
  });

  it('handles onShareAmountChangeHandler in OrdersBasketHeader', () => {
    const stockIdx = 0;
    const shareAmount = '100';
    const wrapper = shallow(<OrdersBasket {...props} />);
    wrapper
      .find(OrdersBasketTable)
      .first()
      .props()
      .onShareAmountChangeHandler(stockIdx, shareAmount);
    expect(props.updateStockOrderShareAmountAction).toHaveBeenCalled();
    expect(props.updateStockOrderShareAmountAction).toHaveBeenCalledWith(
      stockIdx,
      Number(shareAmount)
    );
    restoreMock();
  });

  it('triggers updateStockOrderStatusAction to Ready', () => {
    const p = { ...props, orders: [{ ...props.orders[0], shareAmount: 100 }] };
    const stockIdx = 0;
    const modeId = 0;
    const wrapper = shallow(<OrdersBasket {...p} />);
    wrapper
      .find(OrdersBasketTable)
      .first()
      .props()
      .onExecutionModeSelectHandler(stockIdx, modeId);
    expect(props.updateStockOrderExecutionModeAction).toHaveBeenCalled();
    expect(props.updateStockOrderExecutionModeAction).toHaveBeenCalledWith(
      stockIdx,
      'Market'
    );
    expect(props.updateStockOrderOrderPriceAction).toHaveBeenCalled();
    expect(props.updateStockOrderOrderPriceAction).toHaveBeenCalledWith(
      stockIdx,
      0
    );
    expect(props.updateStockOrderStatusAction).toHaveBeenCalled();
    expect(props.updateStockOrderStatusAction).toHaveBeenCalledWith(
      stockIdx,
      'Ready'
    );
    restoreMock();
  });

  it('triggers updateStockOrderStatusAction to Not Ready', () => {
    const p = {
      ...props,
      orders: [
        {
          ...props.orders[0],
          shareAmount: 100,
          executionMode: 'Market' as OrderExecutionMode,
          status: 'Ready' as OrderStatus,
        },
      ],
    };
    const stockIdx = 0;
    const modeId = 1;
    const wrapper = shallow(<OrdersBasket {...p} />);
    wrapper
      .find(OrdersBasketTable)
      .first()
      .props()
      .onExecutionModeSelectHandler(stockIdx, modeId);
    expect(props.updateStockOrderExecutionModeAction).toHaveBeenCalled();
    expect(props.updateStockOrderExecutionModeAction).toHaveBeenCalledWith(
      stockIdx,
      'Limit'
    );
    expect(props.updateStockOrderStatusAction).toHaveBeenCalled();
    expect(props.updateStockOrderStatusAction).toHaveBeenCalledWith(
      stockIdx,
      'Not Ready'
    );
    restoreMock();
  });

  it('handles onCancelButtonClickHandler in OrdersBasketHeader', () => {
    const stockIdx = 0;
    const wrapper = shallow(<OrdersBasket {...props} />);
    wrapper
      .find(OrdersBasketTable)
      .first()
      .props()
      .onCancelButtonClickHandler(stockIdx);
    expect(props.acknowledgeStockOrderErrorAction).toHaveBeenCalled();
    expect(props.acknowledgeStockOrderErrorAction).toHaveBeenCalledWith(
      stockIdx
    );
    restoreMock();
  });

  it('handles no OrdersBasketTable when there is no orders', () => {
    const wrapper = shallow(<OrdersBasket {...props} />);
    wrapper.setProps({ orders: [] });
    expect(wrapper.find(OrdersBasketTable).length).toBe(0);
    restoreMock();
  });

  it('handles sessionTimeout change', () => {
    const { alert } = window;
    // delete window.alert;
    window.alert = jest.fn();
    const wrapper = shallow(<OrdersBasket {...props} />);
    wrapper.setProps({ sessionTimeout: true });
    expect(window.alert).toHaveBeenCalled();
    expect(props.logOutAction).toHaveBeenCalled();
    restoreMock();
    window.alert = alert;
  });
});
