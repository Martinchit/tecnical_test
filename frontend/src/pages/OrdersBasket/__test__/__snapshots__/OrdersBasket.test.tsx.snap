// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersBasket renders correctly 1`] = `
<styled.div>
  <Component
    title="Orders Basket"
  />
  <Component
    bookButtonDisable={true}
    onBookButtonClickHandler={[Function]}
    onRemoveButtonClickHandler={[Function]}
    removeButtonDisable={true}
  />
  <styled.div
    margin="2% 0%"
  >
    <Component
      onCancelButtonClickHandler={[Function]}
      onExecutionModeSelectHandler={[Function]}
      onOrderPriceChangeHandler={[Function]}
      onSelectCheckClickHandler={[Function]}
      onShareAmountChangeHandler={[Function]}
      orders={
        Array [
          Object {
            "bloombergTickerLocal": "testTicker",
            "currency": "HKD",
            "id": 1,
            "orderId": "9cc4ca45-b42b-462d-a8aa-931e6041cfcb",
            "selected": false,
            "side": "Buy",
            "status": "Not Ready",
            "stockId": "testID",
          },
        ]
      }
    />
  </styled.div>
</styled.div>
`;
